pysph/cpy/parallel.py

```diff
@@ -107,6 +106,256 @@
     return c_${name}(${py_args})
 '''
 
+scan_cy_template = '''
+from cython.parallel import parallel, prange, threadid
+from libc.stdlib cimport abort, malloc, free
+cimport openmp
+cimport numpy as np
+cpdef int get_number_of_threads():
+    % if openmp:
+        cdef int i, n
+        with nogil, parallel():
+            for i in prange(1):
+                n = openmp.omp_get_num_threads()
+        return n
+    % else:
+        return 1
+    % endif
+
+cdef int gcd(int a, int b):
+    while b != 0:
+        a, b = b, a % b
+    return a
+
+cdef int get_stride(int sz, int itemsize):
+    return sz / gcd(sz, itemsize)
+
+
+cdef void c_${name}(${c_arg_sig}):
+    cdef int i, n_thread, tid, stride, sz, N
+
+    N = SIZE
+    n_thread = get_number_of_threads()
+    sz = sizeof(${type})
+
+    # This striding is to do 64 bit alignment to prevent false sharing.
+    stride = get_stride(64, sz)
+
+    cdef ${type}* buffer
+    buffer = <${type}*> malloc(n_thread * stride * sz)
+
+    % if use_segment:
+    cdef int* scan_seg_flags
+    cdef int* chunk_new_segment
+    scan_seg_flags = <int*> malloc(SIZE * sizeof(int))
+    chunk_new_segment = <int*> malloc(n_thread * stride * sizeof(int))
+    % endif
+
+    % if complex_map:
+    cdef ${type}* map_output
+    map_output = <${type}*> malloc(SIZE * sz)
+    % endif
+
+    if buffer == NULL:
+        abort()
```

Comment:
I don't like this, we should raise an exception with a memory error or so if we cannot allocate memory and not just abort, that would be a nasty way to exit.